javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,293 @@
 package java.util
 
-abstract class AbstractList[E] extends AbstractCollection[E] with List[E]
+import scala.collection.JavaConversions._
+
+import scala.annotation.tailrec
+
+abstract class AbstractList[E] protected () extends AbstractCollection[E] with List[E] {
+  self =>
+
+  override def add(element: E): Boolean = {
+    add(size, element)
+    true
+  }
+
+  def add(index: Int, element: E): Unit =
+    throw new UnsupportedOperationException
+
+  def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+    checkIndexOnBounds(index)
+    for ((elem, i) <- c.iterator().zipWithIndex)
+      add(index + i, elem)
+    c.nonEmpty
+  }
+
+  override def clear(): Unit =
+    removeRange(0, size)
+
+  override def equals(o: Any): Boolean = {
+    if (o.asInstanceOf[AnyRef] eq this)
+      return true
+
+    o match {
+      case other: List[_] =>
+        val (e1, e2) = (this.listIterator, other.listIterator)
+        while (e1.hasNext && e2.hasNext) {
+          val (o1, o2 )= (e1.next(), e2.next())
+          if (!(o1 == null && o2 == null) && o1 != o2)
+            return false
+        }
+        !e1.hasNext && !e2.hasNext
+      case _ => false
+    }
+  }
+
+  override def hashCode(): Int = {
+    this.foldLeft(1) {
+      case (hashCode, null) => 31 * hashCode
+      case (hashCode, elem) => 31 * hashCode + elem.hashCode
+    }
```

Comment:
Given the duplication in the two cases, it seems it would be nicer to just have one case and an `if (elem == null) 0 else elem.hashCode`.

Also, consider renaming `hashCode` to `prev`, which is the name we typically use in this codebase for the accumulator of a fold.